### PR TITLE
Distributed Adam optimizer overlaps param all-gather with forward compute

### PR DESCRIPTION
**Pull request by timmoon10**
# What does this PR do ?

This PR adds experimental support in GPT for the distributed Adam optimizer to overlap its parameter all-gather with the forward compute. It integrates with features added in https://github.com/NVIDIA/apex/pull/1559.

This is done with a hacky monkey-patching approach, so long-term it may be better to transition toward PyTorch's FSDP. That said, I have seen some modest speedups (~2%) training GPT-3 175B on Selene with no difference in loss values.

**Collection**: NLP

# Changelog 
- Configure dist Adam so all model parameters are views into a large contiguous buffer.
- Add GPT support for overlapping forward compute and param all-gathers

# Usage

Set the optimizer to `distributed_fused_adam` in the config file:

https://github.com/NVIDIA/NeMo/blob/80cd1c135f4cf01aa04e62aa4c60fb130c889831/examples/nlp/language_modeling/conf/megatron_gpt_config.yaml#L193

The overlapped param all-gathers are enabled by default. To disable, configure the optimizer with `overlap_param_sync: False`.

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [x] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Depends on https://github.com/NVIDIA/NeMo/pull/5560
* Depends on https://github.com/NVIDIA/apex/pull/1559, which in turn depends on https://github.com/NVIDIA/apex/pull/1559.

### CODE REVIEW

**Comment by okuchaiev** [comment]
@timmoon10 can you please resolve merge conflicts

**Review by ericharper — APPROVED** [review]
LGTM. Thanks!
